### ISSUE TITLE
Zoltan2: should the zoltan2 adaptor return a pointer or a view

### ISSUE THREAD

**Issue by bathmatt** [trilinos/Trilinos#9452]
## Enhancement
@trilinos/zoltan2 

Right now you have this code chunk in Zoltan2_Adaptor.hpp

```
  /*! \brief Provide a pointer to this process' identifiers.                                                                                                                      
      \param ids will on return point to the list of the global Ids for                                                                                                           
        this process.                                                                                                                                                             
   */
  virtual void getIDsView(const gno_t *&ids) const {
    // If adapter does not define getIDsView, getIDsKokkosView is called.                                                                                                         
    // If adapter does not define getIDsKokkosView, getIDsView is called.                                                                                                         
    // Allows forward and backwards compatibility.                                                                                                                                
    Kokkos::View<const gno_t *, typename node_t::device_type> kokkosIds;
    getIDsKokkosView(kokkosIds);
    ids = kokkosIds.data();
  }
```

as you can see that this is returning a pointer to the ids vector.  This is not the best idea, specifically since kokkosIds is a temp that is returned from the previous call.  It can go out of scope and have the data returned be bad.  

We should return a view and let the calling code grab a pointer at the very least.  

@kddevin will take the action item on looking into this a bit more and suggesting a better path forward for the adaptors and we will have a path forward on this.

**Comment by kddevin** [comment]
Note:  PR #9441 was accepted, but some of its changes should be reverted once this redesign is complete.  Specifically, the original intent of the Kokkos-based methods in adapters was that users could implement either the pointer-based method OR the Kokkos-based methods, but not necessarily both.  The default methods that converted between pointers and Kokkos (Zoltan2::Adapter's getIDsView and getIDsKokkosView) were not clearly compatible with UVM removal, so we accepted #9441.  Cleaner interfaces as mentioned in this Issue should consider the methods that the user must implement and possibilities to convert between them.

Additionally, Zoltan2  models should be refactored to NOT store pointers to user's views obtained through adapters.  Rather, to the extent possible, the model construction should call Zoltan2 adapters on demand.
